teamcity/response_decoder.go

```diff
@@ -0,0 +1,22 @@
+package teamcity
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+)
+
+type responseDecoder struct{}
+
+func (responseDecoder) Decode(resp *http.Response, v interface{}) error {
```

Comment:
Looking more closely, I don't think this will solve the issue of non-JSON responses. One option is that if unmarshalling fails, don't just return the error, but return an error that also includes the response body in it.

For example:

```
bodyBytes, err := ioutil.ReadAll(resp.Body)
if err != nil {
	return err
}

err = json.Unmarshal(bodyBytes, v)
if err != nil {
	return fmt.Errorf("%v: %s", err, string(bodyBytes))
}
return nil
```

Reply:
Added comment to describe response decoder. Added body in the error message. Unfortunately, for response message to show up in the interface we would need to to Marshal and Unmarshal. Simply unmarshal does not work. using v= also does not work. Both of these results in an empty interface. Unmarshal works only after Marshal.

Reply:
https://play.golang.org/p/as9cn7DN5r1

Reply:
When v is a string, then we don't need to unmarshal at all. We can detect this using type assertion: https://play.golang.org/p/GXCzUsTZO4z

Reply:
done